module-apps/apps-common/popups/VolumeWindow.cpp

```diff
@@ -55,21 +55,41 @@ namespace gui
     {
         WindowWithTimer::onBeforeShow(mode, data);
         const auto popupData = dynamic_cast<VolumePopupData *>(data);
-        if (popupData) {
+        if (popupData != nullptr) {
             volume       = popupData->getVolume();
             audioContext = popupData->getAudioContext();
+            source       = popupData->getVolumeChangeRequestSource();
             if (volumeBar != nullptr) {
                 volumeBar->setValue(volume);
             }
             if (volumeText != nullptr) {
-                showProperText(audioContext, volume);
+                showProperText(audioContext, volume, source);
             }
         }
     }
 
-    void VolumeWindow::showProperText(const audio::Context &audioContext, const audio::Volume volume) noexcept
+    void VolumeWindow::showProperText(const audio::Context &audioContext,
+                                      const audio::Volume volume,
+                                      const audio::VolumeChangeRequestSource source) noexcept
     {
-        volumeText->setText(utils::translate(style::window::volume::base_title_key));
+        if (volume == 0) {
+            showMuted();
+            return;
+        }
+
+        switch (source) {
+        case audio::VolumeChangeRequestSource::A2DP:
+            showBTAudioDeviceVolume();
+            return;
+        case audio::VolumeChangeRequestSource::HFP:
+            [[fallthrough]];
+        case audio::VolumeChangeRequestSource::HSP:
+            showCalling();
+            return;
```

Comment:
AFAIK `[[fallthrough]]` should be used if you do some operations in `case` and then you want to intentionally go to the next `case`. If `case` is empty you don't need to use `[[fallthrough]]`.

Reply:
Related link https://en.cppreference.com/w/cpp/language/attributes/fallthrough